\startchapter{A kinetic model and API of PDI}
\label{PDIpy_chapter}

\section{Introduction}
Antibiotic resistant infections are projected to exceed cancer in annual deaths, and globally cost $10^{13}~USD$ in lost economic production, by mid-21st century \cite{ONeill2014AntimicrobialNations}. Methicillin-resistant \textit{Staphylococcus aureus} (MRSA) \cite{Song2011SpreadStudy,Borg2007PrevalenceCountries} and fluoroquinolone-resistant \textit{Salmonella} \cite{Moghnieh2018EpidemiologyLeague} are two worrisome examples of virulent pathogens that are developing resistance to the antibiotics that subdued them during the 20th century. Antimicrobial resistance (AMR) evolution can be slowed by reducing excessive and incomplete use of antibiotics for human illness and animal agriculture (which is globally the primary consumer of antibiotics \cite{VanBoeckel2017ReducingAnimals,Eggleton2020TheWorld}); however, the fundamental cause of AMR is the conventional strategy of targeting a specific microbial vulnerability. This high selectivity can mitigate off-target effects, however, it places evolutionary pressure on the pathogen to fortify the targeted biochemical vulnerability and thus become immune to the treatement. This arms race of medicinal chemists against microbial evolution must therefore be replaced with a more efficient and sustainable medical strategy, and one that ideally also avoids the persistent ecotoxicity  \cite{Thomas2001AntifoulingEffects,Niu2016RolesIrradiation,Winters1983ControlDesalination}.

\subsection{Photodynamic inactivation}

Photodynamic inactivation (PDI), which differs from photodynamic therapy (PDT) for cancer treatment \cite{Lange2019ComparisonLines} only in its application, offers an effective alternative for treating prokaryotic \cite{Hamblin2004PhotodynamicDisease} and viral \cite{Wigginton2010OxidationInactivation,Lebedeva2020TheViruses} pathogens. PDI describes a photochemical process where reactive oxygen species (ROSs) \cite{Zepp1992HydroxylReaction,Koppenol2001TheLater}, primarily singlet state oxygen ($\ce{^1O2}$, the lowest excitation state of diatomic oxygen) \cite{Ergaieg2008InvolvementPorphyrin, Allen2004IntroductionSimulations, Henze2019Multi-scaleCheckpoint, Zaman2005ComputationalMatrices,Gillespie2007StochasticKinetics}, non-selectively oxidize biological substrate \cite{Choe2006MechanismsOxidation,Frankel1980LipidOxidation}. This mechanism enables PDI to simultaneously 1) avoid resistance evolution \cite{Tavares2010AntimicrobialTreatment,Lauro2002PhotoinactivationConjugates,Pedigo2009AbsenceTherapy}; 2) treat recalcitrant biofilms \cite{Beirao2014PhotodynamicPorphyrin,Ghorbanzadeh2020ModulationModel}, where the biofilm matrix is itself oxidized by $\ce{^1O2}$; and 3) avoid ecological presistence, since $\ce{^1O2}$ has only a $\approx 100 nm$ diffusion distance and a $\approx 10^{-6} s$ lifetime \cite{Moan1984TheOxygen, Moan1990OnTissues,Rodgers1982LifetimeMeasurements} in aqueous environments. The last quality encourages the use of PDI in wastewater treatment \cite{Kohn2007AssociationOxygen,Mostafa2013SingletMatter,Jimenez-Hernandez2006SolarSensitizers} and surfaces \cite{McCoy2014PhotodynamicControl} or industrial polymers \cite{Kim2003DesignProblem} where $\ce{^1O2}$ won't leach into the environment or human consumables. 

The distinction between $\ce{^1O2}$ and triplet state oxygen ($\ce{^3O2}$, the ground conformation of diatomic oxygen) is explained by their quantum multiplicities. The singlet state, of any molecule, contains only paired electrons (couples of electrons with up and down spins) and is named after its multiplicity of $1$: from 
\begin{equation} \label{multiplicity}
    multiplicity = 2(S)+1
\end{equation}
when $S=0$. The $S$ variable in \cref{multiplicity} describes the total angular momentum of the molecule -- the sum of electron spins, where up is $+\frac{1}{2}$ and down is $-\frac{1}{2}$ -- which in the case of a singlet molecule is 0 since the complete pairing of electrons necessitates that the quantities of up and down electrons are equivalent (Figure S1). The molecular triplet state, in contrast, contains two unpaired (radical) electrons that result in a multiplicity of $3$ from $S=1$ in \cref{multiplicity}. These unpaired electrons in $\ce{^3O2}$ increase shielding of the nuclear charges \cite{Katriel1972ARule} and consequently stabilize $\ce{^3O2}$ by $0.98$ eV \cite{Jockusch2008SingletExcitation} relative to $\ce{^1O2}$ that contains lacks this shielding.

The excitation of $\ce{^3O2}$ to $\ce{^1O2}$ is controlled and augmented in PDI by photosensitizer catalysts (PSs). The PS advantageously 1) is a means of localizing oxidation, 2) is a means of controlling the timing and magnitude of excitation; and 3) is a means of generating antimicrobial concentrations of $\ce{^1O2}$ that would not occur by direct excitation $\ce{^3O2 ->[{hv}] ^1O2}$ \cite{Krasnovsky2012PhotochemicalEnvironment}, since the formal selection rules \cite{Bowen1936ForbiddenLines} describe that likely excitations are those which preserve the electronic state: i.e. ground triplet to excited triplet. The $\ce{^3O2}$ could potentially excite to another triplet state and then relax into $\ce{^1O2}$ \cite{Long2003SelectionOxygen}; nevertheless, the PS catalyst accelerates $\ce{^3O2}$ excitation through energy transfers \cite{You2018ChemicalOxygen,Schalk2008Near-infraredTetratolyl-porphyrins,Jockusch2008SingletExcitation}.

The first mechanistic step of PDI describes the ground-state PS ($\ce{^1PS}$) absorbing a photon ($hv$) and entering an excited singlet state ($\ce{^1PS^*}$), according to the selection rules. This excited state then relaxes through intersystem crossing, instead of fluorescing \cite{Kessel1982DeterminantsSpectra}, to an excited triplet state ($\ce{^3PS}$),
\begin{equation} \label{ps_excitation_steps}
    \ce{^1PS <=>[{excitation}][{fluorescence}] ^1PS^* ->[][{intersystem-crossing}] ^3PS}~.
\end{equation}
The second step describes relxation via an energy transfer from $\ce{^3PS}$ to $\ce{^3O2}$, instead of phosphorescence \cite{Mcrae1958Enhancement6}, and thereby generates $\ce{^1O2}$ while regenerating the ground-state $\ce{^1PS}$ catalyst,
\begin{equation} \label{excited_ps_steps}
    \ce{^1 PS <-[{phosphorescence}] ^3 PS ->[^3O2] ^1 PS + ^1O2}
\end{equation}
The combined efficiency of \cref{ps_excitation_steps,excited_ps_steps} is encapsulated in a quantum yield of $\ce{^1O2}$ production \cite{Bakalova2004QuantumPhotosensitizers} ($0\le \Phi_{\ce{^1O2}}\le 1 ~;~ \frac{\ce{^1O2} ~molecules ~produced}{photon ~absorbed}$). The $\ce{^3PS}$ and $\ce{^1O2}$ excited states engage in energy transfers instead of $\ce{^1PS^*}$ and $\ce{^3O2^*}$ since they have longer lifetimes as a consequence of fluorescence being more favorable than phosphorescence. The third step and final of PDI describes $\ce{^1O2}$ reacting with biological substrates through Type II oxidation mechanisms, which are concerted Schenck \cite{Prein1996TheApplications} or Alder-ene \cite{Fernandez-torquemada2012DispersionPlants} reactions that produce organic peroxides \cite{Foote1965ChemistrySelectivity}, as opposed to Type I mechanisms \cite{Bolland1949KineticsOxidation,Farmer1943TheRubber,Grynova2011RevisingAutooxidation} that only affect radical substrates \cite{Litwinienko1999DifferentialEsters}. The Type II mechanism most readily affects allylic positions of unsaturated molecules \cite{Ellison1996ThermochemistryIons,Sehon1950TheRadical}, although, saturated molecules, like those in bacterial phospholipids \cite{ODonnell1985NumericalStaphylococci}, are also oxidized by $\ce{^1O2}$,. 

\begin{figure}[t]
    \centering
    \includegraphics[width = \textwidth]{images/PDIpy/background/jablonski_diagram.png}
    \caption{
        A qualitative Jablonski energy diagram of photosensitization. The initial electronic absorption of a photon ($h\nu$) by $\ce{^3O2}$ forms a $\ce{^3O2}$* molecule, following the selection rules of excitation, which is followed with either fluorescence relaxation or an intersystem-crossing relaxation to form the reactive $\ce{^1O2}$. The singlet $\ce{^1O2}$ molecule either relaxes through phosphorescence or it reacts with an organic substrate to form a peroxide, like the illustrated hydroperoxide with a generic “R” organic group. 
    }
    \label{jablonski_diagram}
\end{figure}

\subsubsection{Photosensitizer}
The $\Phi_{\ce{^1O2}}$ efficiency is primarily dependent upon the chemical structure of the photosensitizer catalyst (PS), notwithstanding minor influence of the chemical conditions \cite{Kruk1998PhotophysicsLuminescence,Kullmann2012UltrafastBisporphyrin}. Two primary inefficiencies of PSs are its propensity to relax through fluorescence or phosphorescence, in Figure \ref{jablonski_diagram}, and to photobleach, where irradiation irreversibly compromises molecular absorptivity \cite{Bonnett2010ChemInformTherapy,Wasser1973TheMetallochlorins}. The functionality and charge of the PS should furthermore optimize its association with the targeted cells \cite{VanDerWal1997DeterminationBacteria,Dickson1989CellSurfaces} and minimize off-target oxidation \cite{Lambrechts2005PhotodynamicMice} and host toxicities \cite{Quishida2016PhotodynamicLight} in medical applications. Material applications of PDI \cite{Peddinti2018PhotodynamicThreat,Gottenbos2001AntimicrobialBacteria}, for applications in either biofouling or sterile medical surfaces, further require that the PS is amenable to permanent surface attachment in a manner that retains material properties \cite{McCoy2014PhotodynamicControl}. 

The PS finally influences the biological targets of PDI. Impermeable PSs, which cannot penetrate a cell, generally oxidize the cytoplasmic membrane \cite{Specht1990DepolarizationAction,Ehrenberg1993ElectricAlterations} instead of cytoplasmic contents \cite{Maisch2004AntibacterialDermatology}. This mechanism, which primarily affects the phospholipid fatty acids in Figure \ref{schenck_mechanism}, manifests in cell death through lysis \cite{Sahu2009AtomicColi,Bertoloni1987RoleCells} and generally affects gram-positive bacteria more than gram-negative bacteria \cite{Lauro2002PhotoinactivationConjugates,Merchat1996Meso-substitutedBacteria} since the latter possess a superficial lipopolysaccharide layer that protects the cytoplasmic membrane. Permeable PSs, by contrast, can penetrate a cell and thus generate $\ce{^1O2}$ within the cytoplasm where cytoplasmic chemicals \cite{Bagchi1979RoleAcriflavine} such as guanine nucleotides \cite{Prat1997Determination9,Devasagayam1991FormationOxygen} are fatally oxidized. This mechanism is more effective with prokaryotes than eukaryotes \cite{Quishida2016PhotodynamicLight}, since the latter have a nuclear membrane that protects DNA, particularly guanine, from oxidation \cite{Pereira2013PhotodynamicVitro}.

A narrow range of chemicals meet these criteria of an ideal PS. Semiconductors \cite{Nelson2002PhotoconductivityDioxide,Peiro2006PhotochemicalPreparations,Linsebigler1995PhotocatalysisResults}, and some amino acid residues \cite{Lippincott-schwartz2003PhotobleachingTechniques,Jin1995PhotolysisSolution}, can electrocatalytically generate $\ce{^1O2}$; however, these molecules are inefficient and/or impractical, particularly for medical applications. The most efficacious PS in nature is chlorophyll \cite{Ramel2012ChemicalPlants}, which is an organometallic porphyrinoid (Figure \ref{zinc_porphyrin}) that evolution has tuned for low rates of photobleaching and absorption of visible light -- specifically blue-violet radiation \cite{Mtangi2017ControlSplitting} via the Soret absorption band \cite{Carre1999FungicidalCerevisiae,Pereira2014InfluencePorphyrin,Ashkenazi2003PhotodynamicBacteria,Moan1986PorphyrinShGroups,Nitzan1992InactivationPorphyrins,Durantini2006PhotodynamicBacteria,Salmon-Divon2004MechanisticTetra-mesoN-methylpyridylporphine} and green-orange radiation \cite{Bertoloni2000PhotosensitizingCells} via the Q absorption band \cite{Bonnett1999PhotobleachingStudy,Jori2006PhotodynamicApplications,Gad2004TargetedMice,Zhao2019Porphyrin-basedAbsorption}. Chlorophyll, however, has not evolved traits that optimize its association with cellular targets or its compatibility with material surfaces; therefore, synthetic porphyrins \cite{Orenstein1997TheInfections,Beirao2014PhotodynamicPorphyrin,Merchat1996StudiesPorphyrins} that emulate the successful conjugated structure \cite{Huang2008Porphyrin-dithienothiopheneCells} of chlorophyll, while introducing other metal centers \cite{Mosinger1997QuantumPorphine} and functional handles  \cite{Hirao1999TheoreticalDerivatives,Wu2014BODIPY-basedSolution,Chacon1988SingletArachidonic} (e.g. Figure \ref{zinc_porphyrin}) that improve its utility in PDI \cite{Jager2016QScales,Karolczak2004PhotophysicalTetraphenylporphyrin,Mathai2007SingletTherapy}, is an appealing direction for PDI research. 

\begin{figure}
    \centering
    \includegraphics[width = \textwidth]{images/PDIpy/background/chlorophyll.png} \\ \midrule
    \includegraphics[width = 0.5\textwidth]{images/PDIpy/background/zinc_porphyrin.png}
    \caption{
        The chemical structure of porphyrinoid chlorophyll (top) juxtaposed with the core motif of a synthetic porphyrin analogue (bottom).
    }
    \label{zinc_porphyrin}
\end{figure}

\begin{figure}[t]
    \centering
    \includegraphics[width =0.9 \textwidth]{images/PDIpy/background/BCFA_schenck_oxidation_2.png}
    \caption{
         The Schenck reaction and associated byproduct decompositions. Step (1) depicts the concerted\cite{Foote1968PhotosensitizedOxygen} Schenck reaction. Step (2) depicts the homolytic cleavage of the hydroperoxide bond to form $\ce{OH^.}$ and an oxy radical that may enter autoxidation mechanisms. Step (3) depicts radical propagation via hydrogen abstraction to form another radical substrate and an alcohol byproduct as part of the autoxidation mechanism. Step (4) is a concerted Russell reaction\cite{Russell1957Deuterium-isotopeRadicals,Howard1968TheMechanism} between two hydroperoxides that forms a $\ce{H2O2}$, an $\alpha,\beta$-ketone, and an alcohol. The reactions of Steps (2-4) sample the wide array of possible oxidative decompositions that follow the Schenck or autoxidation mechanisms.
    }
    \label{schenck_mechanism}
\end{figure}

\subsection{PDI modeling}
Computational models of PDI that allow experimentalists, biologists and chemists alike, to explore the efficacies of different PSs and system conditions are scarce. Santos et al. \cite{Santos2020ApplicationAureus} developed a second-order polynomial to mathematically describe the inactivation of \textit{S. aureus} as a function of time at a particular wavelength and PS concentration. The predictions were demonstrated to be accurate, however, the model is bound to a narrow range of conditions, and does not permit the investigator to explore different parameters such as PS characteristics. Brasel et al. \cite{Brasel2020AnAgalactiae} developed a sigmoidal logistic model to assess the sensitivity of PDI inactivation to incident irradiance $\frac{mW}{cm^2}$ and exposure time; however, the model likewise does not permit investigations of alternative PDI systems. Finally, Sabino et al. \cite{Sabino2019InactivationTherapy} developed a Weibull power-law function from fitted inactivation data to practically predict lethal doses and the tolerance factor of a PDI system; yet, the mathematical model is limited in scope and does not describe the fundamental chemistry of PDI. 

We therefore developed a mechanistically resolved kinetic model of PDI for an impermeable PS and encapsulated into a Python API (PDIpy) that allow investigators to efficiently explore a continuum of values for numerous parameters. The means of editing and customizing the extensive list of parameters, and understanding the default values, is detailed in the API documentation. PDIpy uses Tellurium \cite{Choi2018Tellurium:Biology} to concisely construct SBML \cite{Keating2020Models}, SED-ML \cite{Waltemath2011ReproducibleLanguage}, and COMBINE OMEX \cite{Bergmann2014COMBINEProject} descriptions of the simulations and their results. These conventional formats in computational biology support transparency and reproducibility of the simulation results. The logistic (sigmoidal) Hill equation is then fitted to simulation predictions of cytoplasmic oxidation to systematically construct the inactivation plot based upon the oxidation predictions. We exemplify the model through replicating experimental studies and conducting a sensitivity analyses of key API parameters. We expect that the open-source project will foster experimental progress without expending resources, and will inspire computational biologists to refine tools for this field of medical research, towards expediting the scientific response to the looming medical crisis of antibiotic resistance. 

\section{Methods}
\subsection{Conceptual model}
PDIpy conceptually represents an experimental PDI system with a coccus (spheroid) bacterium like \textit{S. aureus}. Each simulation accepts parameters for each aspect of PDI: the light source and incident intensity; the PS absorptivity, structural dimensions, and $\frac{mol}{vol}$ or $\frac{mol}{area}$ concentration; the bacterial specie, membrane composition, and $\frac{CFU}{mL}$ for planktonic experiments; the solution dimensions; and kinetic constants for the PDI reactions. Default values for many of these parameters supplement user-defined parameters.

\subsection{Kinetic reactions}
Our model is predicated upon a set of three chemical processes that embody the essence of PDI. 1) A photoelectric interaction \cite{Wheaton2009PhotoelectricEffect} from \cref{ps_excitation_steps} occurs after $\ce{^1PS}$ is struck by a photon ($h\nu$) and undergoes intersystem crossing per to form $\ce{^3PS}$. 2) An energy transfer in \cref{excited_ps_steps} occurs from $\ce{^3PS}$ to $\ce{^3O2}$. 3) The $\ce{^1O2}$ oxidizes biological substrates, which includes both cytoplasmic phosholipids and biofilm polymeric substances, or it irreversibly disrupts PS absorptivity through photobleaching 
\begin{equation} \label{bleaching}
    \ce{^1PS -> ^1PS_{bleached}}~.
\end{equation}
A complete description of this kinetic system is represented in Table \ref{reactions_table}, and each of the reactions are each detailed in the following sub-sections.

\subsubsection{User inputs}
The PDIpy API accepts a variety of parameters that allow the user to customize almost every aspect of the PDI system. These parameters of the simulated PDI system can be provided through either a dictionary argument in the \pyobject{define\_conditions} function, or through a JSON parameter file for each category of parameters -- e.g. light, PS, bacterium, and solution space -- that elaborate the simulated system in a more reproducible and transparent manner than dictionary arguments. The complete list of accepted parameters and formats are detailed in the PDIpy documentation (\url{https://github.com/freiburgermsu/PDIpy/blob/main/README.rst}).

\begin{table}[]
    \centering
    \begin{tabular}{c|c}
        \textbf{Name} & \textbf{Reaction} \\
        \toprule
        Photoexcitation & \ce{^1PS <=> ^3PS} \\
        Photobleaching & \ce{^1PS -> ^1PS_{bleached}} \\
        Energy transfer & \ce{^3PS + ^3O2 -> ^1PS + ^1O2} \\
        Phosphorescence & \ce{^1O2 -> ^3O2} \\
        Membrane oxidation & \ce{^1O2 + FA -> oFA} \\
        EPS oxidation & \ce{^1O2 + EPS -> oEPS} \\
    \end{tabular}
    \caption{
        Each of the chemical reactions that define the PDI model of PDIpy. These reactions are individually detailed in dedicated subsections.
    }
    \label{reactions_table}
\end{table}

\subsubsection{Photoelectric}
\paragraph{PS excitation}
PDI begins with the excitation of the PS. This occurs as the combined result of an incident photon i) entering the aqueous solution, ii) striking a PS, and then iii) exciting an electron. This sequence is encapsulated in the kinetic expression
\begin{multline} \label{ps_excitation_kinetics}
    \frac{d[\ce{^3PS}]}{dt} =  k_{excitation}*\frac{photons_{PS}}{photons_{total}} \\ 
    *\Phi_{excitation}*[\ce{^1PS}] - k_{fluorescence}*[\ce{^3PS}]~. 
\end{multline}
The $k_{excitation}$ \& $k_{fluorescence}$ rate constants are estimated as the inverse of the rise and decay times for the selected PS, respectively. The approximate rise time for a porphyrin PS is $50 fs$, which is consistent with estimates of $<100 fs$ \cite{Andersson1999PhotoinducedState} and $60-90 fs$ in ethanol solvent \cite{Gurzadyan1998Time-resolvedZn-tetraphenylporphyrin} which elongates the lifetime of excited molecules relative to water. The decay time, as the S2 fluorescence \cite{Akimoto1999UltrafastPorphyrins}, and $\Phi_{excitation}$ ($\frac{PS excited}{photon absorbed}$) are approximated for a porphyrin PS as $1.5 ns$ and $\approx 0.7$ \cite{Krasnovsky2012PhotochemicalEnvironment}, respectively. The $\frac{photons_{PS}}{photons_{total}}$ \cite{Brasel2020AnAgalactiae}, which is the proportion of photons in the solution that strike a photosensitizer, is calculated through a series of steps. 1) The reported intensity of incident light from the respective light source -- i.e. either irradiance ($\frac{mW}{cm^2}$), lux ($\frac{lumen}{m^2}$), or lumens (lumens) -- is converted into $watts_{incident}$ ($\frac{J}{s}$). 2) This wattage is attenuated by the proportion of the emission spectrum $light_{incident}$ that resides within the excitation spectrum of the parameterized photosensitizer $light_{excitation}$, 
\begin{equation}
    watt_{excitation} = \frac{light_{excitation}}{light_{incident}}*watts_{incident}.
\end{equation}
3) The $watt_{excitation}$ is used to calculate the moles of incident photons that strike photosensitizers per timestep 
\begin{multline} \label{photons_per_second}
    \frac{photons_{PS}}{timestep}=\frac{<h\nu_{excitation}>}{h*c}*watts_{excitation} \\
    *\frac{second}{timestep}*reflection*scattering*\frac{1 mole}{N_A}*\frac{vol_{PS}}{vol_{total}},
\end{multline}
where $reflection \approx 96 \%$ represents the proportion of incident photons that penetrate an aqueous solution \cite{Gross1993SingletLiposomes}; and $scattering$ represents the proportion of light that reaches a specified depth \cite{RobertW.1973TheSea}, which is calculated by 
\begin{equation}
    I_z = I_0*e^{-k*z}~,
\end{equation}
where $I_z$ is the light intensity relative to the $I_0$ after depth $z$, and $k$ is the attenuation coefficient, which is $\approx 0.04~(\frac{1}{m})$ \cite{Lorenzen1972ExtinctionPhytoplankton} for clear waters. The $\frac{vol_{PS}}{vol_{total}}$ represents the volume proportion of the PS ($vol_{PS}$) -- calculated as the product of the quantity of PS molecules and the volume per molecule per its structure -- to the total volume of the solution in which the PS resides ($vol_{total}$) \cite{Santos2020ApplicationAureus}. The average excitation wavelength of the PS ($<h\nu_{excitation}>$) is calculated as the weighted average of the Soret and Q excitation bands, in proportion to their relative contribution in generating $\ce{^1O2}$ \cite{Nitzan2001PhotoinactivationWavelengths,Hoenes2020PhotoinactivationWavelength}, since this averaged wavelength simultaneously embodies contributions of both excitation bands. The resultant $\frac{photons_{PS}}{timestep}$ from \cref{photons_per_second} is then divided by the $\frac{photons_{total}}{timestep}$ to complete the kinetic expression in \cref{ps_excitation_kinetics} that calculates the excitation of PS in each timestep. 

\paragraph{Photobleaching}
The collision of a PS and a photon may alternatively disrupt the conjugated PS system and prevent further excitation, which is depicted in \cref{bleaching}. The kinetic expression 
\begin{equation} \label{bleaching_kinetics}
    \frac{d[\ce{^1PS_{bleached}}]}{dt} = k_{bleaching}*[\ce{^1PS}]*[\ce{^1O2}]~,
\end{equation}
represents this process. The $k_{bleaching}$ rate constant for the oxygen-dependent photobleaching reaction of \cref{bleaching_kinetics} -- as opposed to the oxygen-independent, first-order, photobleaching reaction \cite{Bonnett1999PhotobleachingStudy,Mang1987PhotobleachingTherapy} -- possesses a rate constant of $\approx 600 \frac{cm^2}{J*M}$ \cite{Dysart2005CalculationCells} for porphyrins, which is a function of light exposure $\frac{J}{cm^2}$.

\subsubsection{Energy Transfer}
The energy transfer from $\ce{^3PS}$ to $\ce{^3O2}$  in \cref{excited_ps_steps} is described with by the kinetics expression
\begin{equation} \label{energy_transfer_kinetics}
    \frac{d[\ce{^1O2}]}{dt} = k_{transfer}*\Phi_{transfer}*[^3PS]*[\ce{^3O2}]. 
\end{equation}
The rate constant $k_{transfer}$ is the inverse of the decay time of $\ce{^3PS}$, which for a porphyrin PS appears to be $100 ns$ in aqueous after accounting for the reported value \cite{Kupper2002KineticsOxygen} in acetone solvent that significantly increases the lifetime of excited states \cite{Spikes1992QuantumUroporphyrin}. The $\ce{^1O2}$ phosphorescence side reaction, which often emits a specific infrared wavelength that can be measured to approximate the $[\ce{^1O2}]$ \cite{Macpherson1993DirectCentres}, is kinetically represented 
\begin{equation}
    \frac{d[\ce{^3O2}]}{dt} = k_{phosphorescence}*[\ce{^1O2}]
\end{equation}
where $k_{phosphorescence}$ is a function of $\frac{CFU}{mL}$, since the $\ce{^1O2}$ lifetime is greater in biological material and thus it is proportional with the bacterial concentration in planktonic simulations \cite{Maisch2007TheBacteria}. The minimum lifetime is limited to $3.5\mu s$ for pure water \cite{Baier2005Time-resolvedCells}.

\subsubsection{Oxidation}
The oxidation reactions are the primary consumers of oxygen in the kinetic system. The simulated system, however, is assumed to possess an equilibrium between the solution and the atmosphere, which prevents the depletion of oxygen in the system. This is represented by replacing each oxygen molecule that is consumed during the oxidation of biological substrate in \cref{membrane_oxidation,EPS_oxidation}.

\paragraph{Cytoplasmic membrane} 
The oxidation of cytoplasmic phospholipids, which we approximate as fatty acid chains, is represented as an irreversible second-order reaction \cite{Watabe2007OxidationMembranes.}
\begin{equation} \label{membrane_oxidation}
    \ce{^1O2 + FA -> oFA}.
\end{equation}
and a second-order kinetic expression
\begin{equation} \label{membrane_oxidation_kinetics}
    \frac{d[oFA]}{dt} = k_{fa}*[\ce{^1O2}]*[FA]~.
\end{equation}
The rate constant $k_{fa} \approx 769 \frac{L}{g*s}$ \cite{Mukai2019KineticSolution} and the concentration of fatty acid chains in the cytoplasmic membrane is described in units of $\frac{g}{L}$, after calculating the weighted average MW of fatty acids within the membrane. 

\paragraph{Biofilm matrix} 
The oxidation of EPS is reported to be significant during PDI \cite{Beirao2014PhotodynamicPorphyrin}. This process is represented through an irreversible reaction
\begin{equation} \label{EPS_oxidation}
    \ce{^1O2 + EPS -> oEPS},
\end{equation}
which competes with \cref{membrane_oxidation} for $\ce{^1O2}$ and thereby lessens the efficacy of PDI upon biofilms relative to planktonic organisms. The oxidation of bioiflm polymers is kinetically represented  
\begin{equation} \label{EPS_oxidation_kinetics}
    \frac{d[oEPS]}{dt} = k_{EPS_{oxidation}}*[\ce{^1O2}]
\end{equation}
with an empirical rate constant that we approximated as $37.75~\frac{1}{s}$ for \textit{S. aureus}, and an initial concentration of EPS that is $9x$ greater than the cellular mass, in proportion to mass distributions of biofilms \cite{Flemming2010TheMatrix}.

\subsection{Inactivation fitting}

Simulation results of chemical concentrations over time are processed into predictions of bacterial inactivation through a series of mathematical steps. 1) The proportion of oxidized substrate
\begin{equation} \label{oxidation_proportion}
    ox_{proportion} = \frac{[oFA]}{[oFA]+[FA]}
\end{equation}
is fitted to a sigmoidal curve, similar to other models \cite{Xiong1999AInactivation}. We used the signmoidal Hill-equation \cite{Gesztelyi2012ThePharmacology} as a biochemical kinetic model that derives from mass-action kinetics similar to the Michaelis-Menten kinetic model, however, a fitting module for the Hill-equation did not exist; hence, the HillFit Python module was developed with a variation of the Hill-equation \cite{Inoue2016OscillationActivation} 
\begin{equation} \label{hill_eq}
    y=bottom+\frac{(top-bottom)*x^n}{EC50^n+x^n}~,
\end{equation}
that improves its fit to data, to fit the oxidation data from PDIpy simulations. The regression parameters from the fitted Hill equation are then systematically adjusted to construct an inactivation plot that replicates experimental results while preserving the underlying Hill-equation relationship. These parameter adjustments are presented in Table \ref{hill_parameters} for both simulations of planktonic and biofilm bacteria, since biofilms factors such as impaired diffusion that are not explicitly accounted in our kinetic model and thus must be compensated through other means. The top parameter of \cref{hill_eq} is adjusted asymptotically to a limit that follows an subtly different empirical expression for planktonic $1-10^{-\Omega}$ than biofilm $1-10^{-0.7-\Omega }$ simulations, where $\Omega = wattage^{\frac{1}{5}}-log10(1-final_{oxidation\_proportion})$. This top limit is empirically determined to manifest in the log-inactivation predictions being 1-2 greater than the log-oxidation from \cref{oxidation_proportion}, which embodies our assumption -- and an implicit observation in experimental studies -- that $1-10\%$ of membrane phospholipids must experience oxidation for the cell to lyse. 

\begin{table}[]
    \centering
    \begin{tabular}{l|c|c}
        \textbf{Bacterial state} & \textbf{Hill parameter} & \textbf{Adjustment} \\
        \multirow{2}{}{Planktonic} & EC50 & -76\% \\
         & nH & +100\% \\
         \midrule
         \multirow{2}{}{Biofilm} & EC50 & -65\% \\
         & nH & +120\% \\
    \end{tabular}
    \caption{
        The Hill parameters adjustments that are enacted to create the inactivation plot for both planktonic and biofilm systems. 
    }
    \label{hill_parameters}
\end{table}

% \subsubsection{Cross-linked systems}

% \paragraph{iPDIpy}
% The PDIpy API is graphically interactive through a user interface iPDIpy, which is depicted in Figure \ref{}. This interface provides a convenient means for parameterizing a PDIpy simulation, exporting and importing sets of parameters, parsing the processed simulation data, viewing error messages and calculated values, viewing the output figure of bacterial reduction over time, and parsing the processed data through a build-in function. The iPDIpy version of PDIpy is downloadable from our GitHub repository.

% \begin{figure}
%     \centering
%     \includegraphics{}
%     \caption{
%         The left pane of the GUI will contain command buttons like restarting the simulation or export results, while the right pane will edit the simulation data and or bacterial qualities/species.
%     }
%     \label{iPDIpy_interface}
% \end{figure}


\section{Case Studies}
A PDI experiment from literature that thoroughly described its expeirmental procedure was parameterized in PDIpy to compare the predicted results with the reported results. These comparisons in the following sections all pertain to \textit{S. aureus}, which is a model coccus bacterium that is abundantly described in studies of PDI and AMR infectants as MRSA.

% \subsection{Cross-linked PS}
% Experimental data from our lab with cross-linked PSs (unpublished) was parameterized, in Table \ref{}, into PDIpy. The predicted reduction curve over time in Figure \ref{} agreed with the experimental data at the measured time of 6 hours and 98\% reduction, which supports that the software yields accurate predictions. 

% \paragraph{}
% .

\subsection{Solution PS}

\paragraph{Beirao et al.\cite{Beirao2014PhotodynamicPorphyrin}}
This paper experimentally examines the efficacy of a dissolved PS, over a range of concentrations, against both planktonic and sessile states. The ample experimental details were parameterized into PDIpy, and the predicted log-inactivation was contrasted with the reported log-inactivation at various times, concentrations, and bacterial states in Table \ref{beirao_et_al_data}. The predicted log-oxidation and log-inactivation for a few of the simulated conditions from the study are plotted in Figure \ref{beirao_et_al}, with the respective regression plots from fitting the Hill-equation to the log-oxidation results in Figure \ref{hill_regression}. The very precise fitting of the data -- $R^2 > 0.996$ -- supports that our kinetic model of PDI fundamentally recreates the biochemical relationship that is predicted by the Hill model of kinetics.

\begin{table}[h]
    \centering
    \begin{tabular}{c|c|c|c|c|c}
        Bacterial & [PS] & Inactivation & Reported & Predicted & \multirow{2}{1.2cm}{\%-error}\\
        state & ($\mu m$) & (-log10) & (min) & (min) & \\
        \toprule
        \multirow{3}{1.5cm}{planktonic} & 5 & 7.6 & 51 & 117 & 119\\
        & 10 & 7.6 & 51 & 36 & -29\\
        & 20 & 7.6 & 42 & 12 & -71\\
        \midrule
        \multirow{3}{1.5cm}{sessile} & 5 & 3.6 & 390 & 566 & 45\\
        & 10 & 5 & 390 & 310 & -20\\
        & 20 & 6.3 & 390 & 247 & -37\\
        \bottomrule
    \end{tabular}
    \caption{
        A quantitative comparison of published inactivation data, with both planktonic and sessile, \textit{S. aureus}, with predictions from PDIpy. The \%-error for each prediction is provided in the table as a metric of accuracy
    }
    \label{beirao_et_al_data}
\end{table}

\begin{figure}
    \centering
    \includegraphics[width = 0.9\textwidth]{images/PDIpy/examples/20uM.png}
    \vspace{5mm}
    \midrule
    \vspace{5mm}
    \includegraphics[width = 0.9\textwidth]{images/PDIpy/examples/10uM_biofilm.png}
    \caption{
        Two exemplary figures of PDIpy replications of the Beirao et al. data for a) planktonic and b) biofilm bacterial states.
    }
    \label{beirao_et_al}
\end{figure}

\begin{figure}
    \centering
    \includegraphics[width = 0.9\textwidth]{images/PDIpy/examples/20uM_regression.png}
    \vspace{5mm}
    \midrule
    \vspace{5mm}
    \includegraphics[width = 0.9\textwidth]{images/PDIpy/examples/10uM_biofilm_regression.png}
    \caption{
        The Hill-equation regression of the oxidation plot from Figure \ref{beirao_et_al}a. The high $R^2$ correlation supports that our chemical model of PDI is a sensible recreation of the fundamental biochemistry.
    }
    \label{hill_regression}
\end{figure}


\section{Sensitivity analyses}

\paragraph{Light intensity}
The sensitivity of simulation results to light intensities, across the range of $[10, 100,000] Lux$, was explored. The trend over this 3-log range, which is represented by Figure \ref{light_intensities}, is that the proportion of excited PS asymptotically approches 100\%. The influence of photobleaching appears to be negligible at the examined time length and light intensity, where a negative slope in the plotted excitation proportion would be expected over time as a decreasing proportion of PSs are able to electronically excite. The analysis further clarifies the minmal value of greater irradiation beyond $\approx 13,000 lux$ from the perspective of exciting PSs, which may be an informative for experimentalists when designing PDI systems.

\begin{figure}
    \centering
    \includegraphics[width = 0.8\textwidth]{images/PDIpy/sensitivity_analyses/215_lux.png} \\
    \vspace{5mm}
    \midrule
    \vspace{5mm}
    \includegraphics[width = 0.8\textwidth]{images/PDIpy/sensitivity_analyses/100000_lux.png}
    \caption{
        The proportion of excited PS at two contrasting light intensities: a) $215 Lux$ and b) $100000 Lux$. The negative slope of the plots is the consequence of photobleaching, where the quantity of excitable PSs decreases over time. This effect is more prominent in b) since the light intensity is much greater than a), and since photobleaching is a function of light intensity.
    }
    \label{light_intensities}
\end{figure}

\section{Discussion}
The preliminary examples of PDIpy with the study from Beirao et al. support that the underlying kinetic model sufficiently represents PDI. The excessive breadth in the simulation predictions, where the log-inactivation predictions with low concentrations occurs too late while the log-inactivation predictions with high concentrations occurs too early, may be resolved with adjusted the Hill parameters that are defined in Table \ref{hill_parameters}. The tighter spread of the biofilm simulation relative to the planktonic simulation supports this method of improving the predicted results, where the different set of parameter adjustments may be better tailored to inactivation phenomena. 

The sensitivity analysis of the light intensity revealed that the relationship between the proportion of PS excitation and the light intensity plateaus beyond $\approx 10,000 lux$. This insight and this type of broad inquiry over a range of values cannot be replicated with existing PDI models, since they neither consider the fundamental kinetics that results in PS excitation nor are encapsulated in a dynamic interface like the PDIpy API.  

The open-source implementation of this model in PDIpy will practically support experimentalists as they develop PDI technologies, and will ideally inspire other computational biologists to develop programs that can foster discovery of antibiotic methods at an imperative time to prevent antimicrobial resistant epedemics.   


\section{Author Contributions}
\begin{description}
    \item[APF] Designed, executed, and codified the project.
    \item[JRK] Guidance and manuscript edits.
    \item[HLB] Guidance, manuscript edits, and funding.
\end{description}

\section{Acknowledgments}

The authors are grateful to Ethan Sean Chan for his development of the framework for iPDIpy, which will be introduced in a future release of PDIpy. The authors thank the members of the Buckley and Wolff Groups at the University of Victoria for contributing ideas and data that were used to refine this PDI model. Andrew finally thanks Hiroaki Imoto for his contributions and guidance in developing the HillFit module that was used to fit data in PDIpy.